src/System.Text.Primitives/System/Text/Encoders/Utf8Decoder.cs

```diff
@@ -0,0 +1,264 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System.Diagnostics;
+using System.Runtime.CompilerServices;
+
+namespace System.Text.Primitives.System.Text.Encoders
+{
+    /// <summary>
+    /// Contains facilities for validating and extracting Unicode scalar values from UTF8 byte sequences.
+    /// </summary>
+    internal static class Utf8Decoder
+    {
+        // Assumption: bytesAvailable > 0
+        /// <summary>
+        /// Reads a Unicode scalar value from the provided UTF8 sequence.
+        /// </summary>
+        /// <param name="data">A reference to the first byte of the UTF8 sequence from which to read the scalar value.</param>
+        /// <param name="bytesAvailable">The number of elements available in the buffer referenced by <paramref name="data"/>.</param>
+        /// <param name="bytesConsumed">When this method returns, contains the number of bytes
+        /// from <paramref name="data"/> which were consumed while reading the scalar value.</param>
+        /// <returns>On success, a non-negative integer which represents a Unicode scalar value.
+        /// On failure, a negative integer which represents an <see cref="ErrorStatus"/> failure code.</returns>
+        /// <remarks>
+        /// The caller must use extreme caution to avoid passing a value for <paramref name="bytesAvailable"/> that is
+        /// larger than the actual buffer referenced by <paramref name="data"/>, else a buffer overrun could occur.
+        /// See http://www.unicode.org/glossary/#unicode_scalar_value for the definition of Unicode scalar value.
+        /// </remarks>
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        public static int DangerousReadUnicodeScalarValue(ref byte data, int bytesAvailable, out int bytesConsumed)
```

Comment:
Do you think this API will ever be public? If yes, I worry about returning error codes mixed with return values. It's not super .net like,